src/client-side-metrics/gcp-metrics-handler.ts

```diff
@@ -0,0 +1,313 @@
+   * @param {OnOperationCompleteData} data Data related to the completed operation.
+   */
+  onOperationComplete(data: OnOperationCompleteData) {
+    this.initialize({
+      projectId: data.projectId,
+      instanceId: data.metricsCollectorData.instanceId,
+      table: data.metricsCollectorData.table,
+      cluster: data.metricsCollectorData.cluster,
+      zone: data.metricsCollectorData.zone,
+    });
+    this.otelMetrics?.operationLatencies.record(data.operationLatency, {
+      appProfileId: data.metricsCollectorData.appProfileId,
+      methodName: data.metricsCollectorData.methodName,
+      clientUid: data.metricsCollectorData.clientUid,
+      finalOperationStatus: data.finalOperationStatus,
+      streamingOperation: data.streamingOperation,
+      clientName: data.clientName,
+    });
+    this.otelMetrics?.retryCount.add(data.retryCount, {
+      appProfileId: data.metricsCollectorData.appProfileId,
+      methodName: data.metricsCollectorData.methodName,
+      clientUid: data.metricsCollectorData.clientUid,
+      finalOperationStatus: data.finalOperationStatus,
+      clientName: data.clientName,
+    });
+    this.otelMetrics?.firstResponseLatencies.record(data.firstResponseLatency, {
+      appProfileId: data.metricsCollectorData.appProfileId,
+      methodName: data.metricsCollectorData.methodName,
+      clientUid: data.metricsCollectorData.clientUid,
+      finalOperationStatus: data.finalOperationStatus,
+      clientName: data.clientName,
+    });
+  }
+
+  /**
+   * Records metrics for a completed attempt of a Bigtable operation.
+   * This method records attempt latency, connectivity error count, server latency,
+   * along with the provided attributes.
+   * @param {OnAttemptCompleteData} data Data related to the completed attempt.
+   */
+  onAttemptComplete(data: OnAttemptCompleteData) {
+    this.initialize({
+      projectId: data.projectId,
+      instanceId: data.metricsCollectorData.instanceId,
+      table: data.metricsCollectorData.table,
+      cluster: data.metricsCollectorData.cluster,
+      zone: data.metricsCollectorData.zone,
+    });
+    this.otelMetrics?.attemptLatencies.record(data.attemptLatency, {
+      appProfileId: data.metricsCollectorData.appProfileId,
+      methodName: data.metricsCollectorData.methodName,
+      clientUid: data.metricsCollectorData.clientUid,
+      attemptStatus: data.attemptStatus,
+      streamingOperation: data.streamingOperation,
+      clientName: data.clientName,
+    });
+    this.otelMetrics?.connectivityErrorCount.add(data.connectivityErrorCount, {
+      appProfileId: data.metricsCollectorData.appProfileId,
+      methodName: data.metricsCollectorData.methodName,
+      clientUid: data.metricsCollectorData.clientUid,
+      attemptStatus: data.attemptStatus,
+      clientName: data.clientName,
+    });
+    this.otelMetrics?.serverLatencies.record(data.serverLatency, {
```

Comment:
what happens if serverLatency is null?

Reply:
I looked into this and if server latency is null then the metric simply doesn't make it to the export function.